### ISSUE TITLE
Export to PDF not showing logo (Have checked and tried everything i can find)

### ISSUE THREAD

**Issue by mcsnoogins** [simpleinvoices/simpleinvoices#305]
Just moved install to new VPS (using the transfer tool in WHM to transfer the whole account), everything works apart from the export to PDF (emailed invoices also missing logo since they are PDFs).  Export to word shows logo. Have check GD2 installed and enabled:

GD Support			enabled
GD Version			bundled (2.1.0 compatible)
FreeType Support		enabled
FreeType Linkage		with freetype
FreeType Version		2.4.11
T1Lib Support			enabled
GIF Read Support		enabled
GIF Create Support		enabled
JPEG Support			enabled
libJPEG Version			6b
PNG Support			enabled
libPNG Version			1.5.13
WBMP Support			enabled
XPM Support			enabled
libXpm Version			30411
XBM Support			enabled
WebP Support			enabled
Using PHP 5.6 as 7 broke most things, master version of simple invoices from git hub.
New server is CENT OS 7 with apache 2.4, php 5.4, 5.5,5.6 and 7.1 with 5.6 being used, database is MariaDB 10.3.18.
Old server was CENT OS 6 with everything else identical, install still works on the old server fine but can't get the logo to show in PDF on the new one :( 
Have also ensured the tmp folder and pdf font folder are set to 777.
Not 100% if it's an issue or a problem with my competence but hoping it's not me as everything seems to be the same.

**Comment by mcsnoogins** [comment]
Anyone able to help with this? I've tried the simple invoices website which seems to be redundant now (if the project needs hosting i'd gladly donate) google groups helped me first time round settings things up but it seems to have closed down as well now :( can't find anywhere else to ask.

**Comment by mcsnoogins** [comment]
Have just realised its not just the logo that's missing but the formatting is wrong as well (coloured title bars missing etc) have attached what i'm getting on the new VPS compared to the old
![pdf](https://user-images.githubusercontent.com/56550035/66852398-cad96380-ef74-11e9-8bbc-b9af75b28548.jpg)
![pdf2](https://user-images.githubusercontent.com/56550035/66852400-cad96380-ef74-11e9-93bd-7ce743253d99.jpg)

**Comment by pixel-penguin** [comment]
Did you manage to fix the issue? Also, are you using Cpanel?

**Comment by pixel-penguin** [comment]
I had the same problem.

https://github.com/simpleinvoices/simpleinvoices/issues/303

**Comment by mcsnoogins** [comment]
Yes using cpanel, I'm actually paying a php guy to look into it as my business relies on this to work.  Doesn't look very likely though, apparently the pdf lib isn't accessing external sources, why that is I don't know but he's looking into using something else to generate the pdfs and see if he can integrate that instead. Tbh I can see me abandoning it and using different software.

**Comment by pixel-penguin** [comment]
on php.ini (on cpanel) did you activate these?

![image](https://user-images.githubusercontent.com/8964456/66985172-80fb9500-f0bc-11e9-81e2-5028bbb5c71e.png)

**Comment by pixel-penguin** [comment]
If you keep struggling, I can always help you out.

**Comment by mcsnoogins** [comment]
Yup, I had all of them enabled at one point, although I don't seem to have the allow url setting in mine :/

**Comment by mcsnoogins** [comment]
Actually just found it, it's on already :(

**Comment by jetibest** [comment]
I programmed some fix, basically using a completely different PDF renderer, i.e. using puppeteer:

**simpleinvoices/library/htmltopdf/main.js**:
```js
const fs = require('fs');
const puppeteer = require('puppeteer');
(async () =>
{
    const htmldata = (await fs.promises.readFile(process.argv[2])).toString();
    const browser = await puppeteer.launch({
        headless: true,
        args: ['--no-sandbox']
    });
    const page = await browser.newPage();
    await page.setContent(htmldata);
    await page.emulateMedia('screen');
    await page.pdf({path: process.argv[3], format: 'A4', printBackground: true});
    await browser.close();
})();
```

This depends on NodeJS and npm, which I already had installed on my server, which is why I chose for this specific solution. But you may use any command-line tool that works for you, to convert html to pdf. If you use the above code, make sure to use `npm install puppeteer` inside the directory with `main.js`.

Additionally, in order to make puppeteer run correctly, I had to install these font packages:
```bash
yum install pango.x86_64 libXcomposite.x86_64 libXcursor.x86_64 libXdamage.x86_64 libXext.x86_64 libXi.x86_64 libXtst.x86_64 cups-libs.x86_64 libXScrnSaver.x86_64 libXrandr.x86_64 GConf2.x86_64 alsa-lib.x86_64 atk.x86_64 gtk3.x86_64 ipa-gothic-fonts xorg-x11-fonts-100dpi xorg-x11-fonts-75dpi xorg-x11-utils xorg-x11-fonts-cyrillic xorg-x11-fonts-Type1 xorg-x11-fonts-misc
```

You can try this code using:
```bash
echo '<!DOCTYPE html><html><body><h1>Hello world</h1><p>This is a test.</p></body></html>' > simpleinvoices/test.html
node simpleinvoices/library/htmltopdf/main.js simpleinvoices/test.html simpleinvoices/test.pdf
```
And then check out http://example.com/simpleinvoices/test.pdf.

But in any case, in order to run your custom html-to-pdf converter tool from simpleinvoices.. Find the `case "pdf":` and replace the code as follows:

**simpleinvoices/include/class/export.php**:
```php
// (...)
case "pdf":
{
    // include stylesheet inline
    $lines = explode("\n", $data);
    for($i=0;$i<count($lines);$i++)
    {
        $line = $lines[$i];
        preg_match('/^.*<link.*"[^"]*(invoices[^"]*\.css)".*>.*$/', $line, $matches);
        if($matches AND count($matches) >= 2)
        {
            $lines[$i] = "<style type=\"text/css\">" . file_get_contents("templates/" . $matches[1]) . "</style>";
        }
    }
    $data = implode("\n", $lines);
    file_put_contents('tmp/' . $this->file_name . '.html', $data);
    exec('/usr/bin/node library/htmltopdf/main.js tmp/' . $this->file_name . '.html tmp/' . $this->file_name . '.pdf');
    echo "<!DOCTYPE html><html><body><a target=\"_blank\" download=\"" . $this->file_name . ".pdf\" href=\"tmp/" . $this->file_name . ".pdf\">Download PDF: " . $this->file_name . "</a></body></html>";
    break;
}
// (...)
```

I don't suppose a logo would work, if it is using a local path, because you would need to fix the path to work in puppeteer. You can fix the path in the template file, or just use an absolute URL. So, for example:

**simpleinvoices/templates/invoices/yourtemplatenamehere/template.tpl**:
```html
<img src="{$logo|urlsafe}" (...)>
<!-- replace the above with absolute URL: -->
<img src="https://example.com/logo.png" (...)>
```

It is a bit messy to patch this neatly into simpleinvoices, but it worked for me with least amount of effort.

Note that if you use another renderer, you might need to change your stylesheet, and configure different margins etc. Also fonts may appear a bit differently. As a quick fix, to make it similar to what I had before, I added `html {zoom: 0.6;padding: 100px;}` to the `style.css` file next to the `template.tpl`.

For simpleinvoices developers: it would be good to improve modularity, by calling the html-to-pdf through an exec call like this workaround, so one can easily use another renderer if one does not work on their system.
